include/swift/SIL/Notifications.h

```diff
@@ -13,9 +13,210 @@
+
+  virtual ~DeserializationNotificationHandlerBase() = default;
+};
+
+/// A no-op implementation of DeserializationNotificationHandlerBase. Intended
+/// to allow for users to implement only one of the relevant methods and have
+/// all other methods be no-ops.
+class DeserializationNotificationHandler
+    : public DeserializationNotificationHandlerBase {
+public:
+  /// Observe that we deserialized a function declaration.
+  virtual void didDeserialize(ModuleDecl *mod, SILFunction *fn) override {}
+
+  /// Observe that we successfully deserialized a function body.
+  virtual void didDeserializeFunctionBody(ModuleDecl *mod,
+                                          SILFunction *fn) override {}
+
+  /// Observe that we successfully deserialized a witness table's entries.
+  virtual void didDeserializeWitnessTableEntries(ModuleDecl *mod,
+                                                 SILWitnessTable *wt) override {
+  }
+
+  /// Observe that we successfully deserialized a default witness table's
+  /// entries.
+  virtual void didDeserializeDefaultWitnessTableEntries(
+      ModuleDecl *mod, SILDefaultWitnessTable *wt) override {}
+
+  /// Observe that we deserialized a global variable declaration.
+  virtual void didDeserialize(ModuleDecl *mod,
+                              SILGlobalVariable *var) override {}
+
+  /// Observe that we deserialized a v-table declaration.
+  virtual void didDeserialize(ModuleDecl *mod, SILVTable *vtable) override {}
+
+  /// Observe that we deserialized a witness-table declaration.
+  virtual void didDeserialize(ModuleDecl *mod,
+                              SILWitnessTable *wtable) override {}
+
+  /// Observe that we deserialized a default witness-table declaration.
+  virtual void didDeserialize(ModuleDecl *mod,
+                              SILDefaultWitnessTable *wtable) override {}
+
+  virtual StringRef getName() const = 0;
+
+  virtual ~DeserializationNotificationHandler() = default;
+};
+
+/// A notification handler that only overrides didDeserializeFunctionBody and
+/// calls the passed in function pointer.
+class FunctionBodyDeserializationNotificationHandler final
+    : public DeserializationNotificationHandler {
+public:
+  using Handler = void (*)(ModuleDecl *, SILFunction *);
+
+private:
+  Handler handler;
+
+public:
+  FunctionBodyDeserializationNotificationHandler(Handler handler)
+      : handler(handler) {}
+  virtual ~FunctionBodyDeserializationNotificationHandler() {}
+
+  void didDeserializeFunctionBody(ModuleDecl *mod, SILFunction *fn) override {
+    (*handler)(mod, fn);
+  }
+
+  StringRef getName() const override {
+    return "FunctionBodyDeserializationNotificationHandler";
+  }
+};
+
+/// A type that enables us to both pass in custom notification handler types
+/// that use a unique_ptr as well as pass managers and analyses that we want
+/// to pass in a bare pointer for. We use the API
+class DeserializationNotificationHandlerSet
```

Comment:
"Maintain a set of unique pointers to custom notifications." Not sure what more you want say.